apis/python/src/tiledbsoma/_tdb_handles.py

```diff
@@ -51,36 +51,45 @@ def open(
     timestamp: Optional[OpenTimestamp],
 ) -> "Wrapper[RawHandle]":
     """Determine whether the URI is an array or group, and open it."""
-    obj_type = tiledb.object_type(uri, ctx=context.tiledb_ctx)
-    if not obj_type:
-        raise DoesNotExistError(f"{uri!r} does not exist")
+    open_mode = clib.OpenMode.read if mode == "r" else clib.OpenMode.write
+    timestamp_ms = context._open_timestamp_ms(timestamp)
 
     try:
-        return _open_with_clib_wrapper(uri, mode, context, timestamp)
-    except SOMAError:
+        soma_object = clib.SOMAObject.open(
+            uri, open_mode, context.native_context, (0, timestamp_ms)
+        )
+
+        if open_mode == clib.OpenMode.read and soma_object.type == "SOMADataFrame":
+            return DataFrameWrapper._from_soma_object(soma_object, context)
+
         # This object still uses tiledb-py and must be handled below
-        if obj_type == "array":
+        if soma_object.type in (
+            "SOMADataFrame",
+            "SOMASparseNDArray",
+            "SOMADenseNDArray",
+        ):
             return ArrayWrapper.open(uri, mode, context, timestamp)
-        if obj_type == "group":
+        if soma_object.type in ("SOMACollection", "SOMAExperiment", "SOMAMeasurement"):
             return GroupWrapper.open(uri, mode, context, timestamp)
+    except (SOMAError, RuntimeError):
```

Comment:
I found this code a bit confusing and with some redundancy -- it suggests that there exists a situation where an exception is thrown in the first part (above), which can be resolved by the "use tiledb-py code path" code below.  The actual code in each path seems only different in that it uses `tiledb.object_type` instead of the tiledbsoma `open` to determine object type (and lacks the tiledbsoma DataFrame read path).

It seems like it could be simplified quite a bit, unless I'm missing some not-obvious logic (in which case a comment might help).  Example of what I had in mind (caveat: not tested):

```
    soma_object = clib.SOMAObject.open(
        uri, open_mode, context.native_context, (0, timestamp_ms)
    )

    obj_type = soma_object.type or tiledb.object_type(uri, ctx=context.tiledb_ctx)

    if open_mode == clib.OpenMode.read and obj_type == "SOMADataFrame":
        return DataFrameWrapper._from_soma_object(soma_object, context)

    if obj_type in (
            "SOMADataFrame",
            "SOMASparseNDArray",
            "SOMADenseNDArray",
            "array",
        ):
            return ArrayWrapper.open(uri, mode, context, timestamp)
        if soma_object.type in ("SOMACollection", "SOMAExperiment", "SOMAMeasurement", "group"):
            return GroupWrapper.open(uri, mode, context, timestamp)

    # Invalid object
    raise SOMAError(f"{uri!r} has unknown storage type {obj_type!r}")
```

Reply:
Thanks for the suggestion. I was able to clean up this code block into something easier to read.